Implements hot reload of the KMS `EncryptionConfiguration`

nilekhc opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind feature

#### What this PR does / why we need it:

1. Implements hot reload of `EncryptionConfiguration`

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #111919

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

The reload flag was already release-noted in https://github.com/kubernetes/kubernetes/pull/113529

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3299-kms-v2-improvements
```

nilekhc commented:
/cc @ritazh @aramase

aramase commented:
/priority important-soon
/triage accepted

nilekhc commented:
/test pull-kubernetes-e2e-kind

nilekhc commented:
/test pull-kubernetes-integration

nilekhc commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,258 @@
+	}
+
+	// update transformers.
+	for groupResource, transformer := range transformerOverrides {
+		ec.storageFactory.UpdateMutableTransformer(groupResource, transformer)
+	}
```
This is not enough because it misses the health checks and leaks the old gRPC connections (which you do not notice in tests because they are still cleaned up at shutdown).

nilekhc replied:
implemented closing grpc conn.

enj reviewed (commented):
I will sync with you offline, this will be messy.

enj reviewed (commented):
Initial feedback.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,255 @@
+		}
+	}, time.Minute, ctx.Done())
+
+	<-ctx.Done()
```
This is arguably a nit and not necessarily a hard requirement.

nilekhc replied:
I thought about it, but we are passing stop-ch from Run to the child go routines so there won't be any leak.

enj replied:
Oh I am not concerned about a leak here - I am saying that `.Run` is blocking, and when it returns, I would expect all of its child go routines to have stopped.  But this is good enough for now.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,255 @@
+	go wait.Until(ec.runWorker, time.Second, ctx.Done())
+
+	// start the loop that watches the encryption config file until stopCh is closed.
+	go wait.Until(func() {
+		if err := ec.watchEncryptionConfigFile(ctx.Done()); err != nil {
+			klog.ErrorS(err, "Failed to watch encryption config file, will retry later")
+		}
+	}, time.Minute, ctx.Done())
```
Hmm I see this is based off

https://github.com/kubernetes/kubernetes/blob/deb67123585ef6a251c0716bc3fe50824b21e7eb/staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamic_serving_content.go#L138-L143

which was updated in #104102

Leave this as-is for now.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -94,6 +95,7 @@ func (p *kmsv2PluginProbe) toHealthzCheck(idx int) healthz.HealthChecker {
 	})
 }
 
+// LoadEncryptionConfig loads the encryption configuration from the provided config file.
```
Yeah. That's GH copilot 😬. I'll come up with something meaningful.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -66,9 +69,10 @@ type EtcdOptions struct {
 	transformerOverrides   map[schema.GroupResource]value.Transformer
```
`LoadEncryptionConfig` returns the map which we then convert to Dynamic transformers. So don't we need a map var to store that initial transformer overrides?

So if we convert this to a dynamic transformer it will be like a chicken and egg situation. We want to convert the map to dynamic transformers, but now we can't do it since dynamic transformers need dynamic transformers to initialize.

nilekhc replied:
nevermind. removed it.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,255 @@
+// hasEncryptionConfigChanged returns true if it has changed
+func (ec *DynamicKMSEncryptionConfigContent) hasEncryptionConfigChanged(newEncryptionConfig []byte) bool {
+	currentConfig := ec.encryptionConfig.Load()
+	if currentConfig == nil {
+		return true
+	}
```
Oh yeah. I had this guard before I added `storeInitialConfig`.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,255 @@
+// processEncryptionConfig processes the encryption config.
+func (ec *DynamicKMSEncryptionConfigContent) processEncryptionConfig(encryptionConfig []byte) error {
+	transformerOverrides, _, err := encryptionconfig.LoadEncryptionConfig(ec.filePath, ec.apiServerDrainCh)
+	if err != nil {
+		klog.Warning("KMS encryption configuration is not valid.", err)
+		return err
+	}
```
(This is still outstanding)

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -513,3 +515,89 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+	inFlight             *atomic.Value
+}
+
+// NewDynamicTransformers returns a new DynamicTransformers.
```
Give this a better comment (or remove it) :)

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -513,3 +515,89 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+	d.closeTransformers.Store(closeTransformers)
+	d.inFlight.Store(inFlight)
+
+	if oldCloseTransformers != nil && oldInFlight != nil {
```
There should be no `nil` checks - this should always work.

nilekhc replied:
No.  For the first time, they will be nil. Same explanation as https://github.com/kubernetes/kubernetes/pull/112050#discussion_r999969690

nilekhc replied:
removed nil checks

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -513,3 +515,89 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+}
+
+func (r *resourceTransformer) TransformToStorage(ctx context.Context, data []byte, dataCtx value.Context) ([]byte, error) {
+	inFlight := r.inFlight.Load().(*atomic.Int32)
+	inFlight.Add(1)
+	defer inFlight.Add(-1)
```
Everything but the `defer` should be in the `transformer` func, which needs to be updated to return an error once the transformer is in a closing state.

nilekhc replied:
updated.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -513,3 +515,89 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+	closeTransformers context.CancelFunc,
+	inFlight *atomic.Int32,
+) (context.CancelFunc, *atomic.Int32) {
+	oldCloseTransformers := d.closeTransformers.Load()
+	oldInFlight := d.inFlight.Load()
+
+	// store new values
+	d.transformerOverrides.Store(transformerOverrides)
+	d.closeTransformers.Store(closeTransformers)
+	d.inFlight.Store(inFlight)
```
This needs to be a single automatic swap.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,291 @@
+		return "", err
+	}
+
+	return fmt.Sprintf("%x", hash.Sum(nil)), nil
```
Just use the bytes directly.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,291 @@
+}
+
+func (d *DynamicKMSEncryptionConfigContent) computeHash(data []byte) (string, error) {
+	hash := sha256.New()
+	_, err := hash.Write(data)
+	if err != nil {
+		return "", err
+	}
```
```suggestion
	hash := sha256.Sum256(data)
```

And drop the error return.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,291 @@
+	return encryptionConfig, nil
+}
+
+// storeInitialConfig when api-server first starts.
+func (d *DynamicKMSEncryptionConfigContent) storeInitialConfig() error {
+	encryptionConfig, err := d.readEncryptionFileContent(d.filePath)
+	if err != nil {
+		return err
+	}
+
+	encryptionConfigHash, err := d.computeHash(encryptionConfig)
+	if err != nil {
+		klog.Warningf("failed to compute hash for encryption config file %s: %v", d.filePath, err)
+		return err
+	}
+
+	d.encryptionConfig.Store(encryptionConfigHash)
+
+	return nil
+}
```
This is a very good point. Updated.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,291 @@
+	var inFlight atomic.Int32
+	cancel, oldInFlight := d.dynamicTransformers.Set(transformerOverrides, closeTransformers, &inFlight)
+
+	if oldInFlight != nil && cancel != nil {
```
No. For the very first time when we initialize dynamicTransformers, they will be nil. Because need to `Load` them before the `Store` operation is carried out in the `Set` method and for the first instance, they will always be nil.

enj replied:
Fix the constructor so the nil checks disappear.

nilekhc replied:
We can't. From the constructor, we call the Set method which initializes DT for the first time. Btw. This code is removed from this file and polling is now in the Set method with nil checks.

enj replied:
Prep the atomic values directly in the constructor so that you can call .Set without nil checks.

nilekhc replied:
Done

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -513,3 +515,89 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+	}
+
+	// new inFlight variable to track transformers that will be set in here.
+	var inFlight atomic.Int32
```
We need a signed integer to decrement the counter in TransformFrom/ToStorage methods. Changed this to `Int64`

enj replied:
Technically you could add `MaxUint - 1` but that is a level of crazy we should avoid.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,258 @@
+
+	encryptionConfig.primeController()
+
+	return encryptionConfig, nil
```
This can no longer error.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -513,3 +515,104 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+	closeTransformers context.CancelFunc,
+	inFlight *atomic.Int64,
+) {
+	tracker := d.transformTracker.Load().(*transformTracker)
```
Use `.Swap`, not `.Load`+`.Store`

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -513,3 +515,104 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+			return true, nil
+		}
+		return false, nil
+	}, d.stopCh)
```
Hmm if `oldInFlight` never reaches zero (say the transformer hangs forever), this would poll forever since this stop channel is only closed on process exit.  I think `.Set` should simply mark the tracker as `closing`, and in the `Transform` methods of `resourceTransformer` we should call `oldCloseTransformers` if `inFlight` `==` `0` and `closing==true`.  This removes the need for this `stopCh` and polling entirely.

After `closing`, `.Set` would probably need to check if `inFlight` `==` `0`, and then also call `oldCloseTransformers`.  In the `Transform` methods, if `closing==true` and `inFlight` `==` `0` on start, we would likely need to issue another `Load`.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -513,3 +515,104 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+}
+
+// Set sets the transformer overrides. This method is not go routine safe and must only be called by the same, single caller throughout the lifetime of this object.
+func (d *DynamicTransformers) Set(
```
`.Set` no longer needs to take `inFlight` as input.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,258 @@
+	// dynamicTransformers updates the transformers when encryption config file changes.
+	dynamicTransformers *encryptionconfig.DynamicTransformers
+
+	apiServerDrainCh <-chan struct{}
```
This code shouldn't care what channel is given, the caller is responsible for picking the right one (and I expect over time that will change)

```suggestion
	stopCh <-chan struct{}
```

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -247,16 +253,46 @@ func (s *EtcdOptions) ApplyWithStorageFactoryTo(factory serverstorage.StorageFac
+					// we ignore the cancel func because this context should only be canceled when stopCh is closed
+					ctx, _ := wait.ContextForChannel(hookContext.StopCh)
+
+					kmsConfigController, err := kmsconfigcontroller.NewDynamicKMSEncryptionConfiguration(
```
Hmm lets move this `New` call into `.Complete`, and just have this hook call `.Run`.  We also need to come up with a way to ensure that `.Run` only happens once because `dynamicTransformers` relies on that since it is not safe to have multiple callers to `.Set`

enj replied:
Maybe we would call `.Run` in `.Complete` - that seems wrong because that is what hooks are supposed to be for?  Or set up the hook in there by passing `c.AddPostStartHook` as a param?

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -247,16 +254,46 @@ func (s *EtcdOptions) ApplyWithStorageFactoryTo(factory serverstorage.StorageFac
+						"kms-encryption-config",
+						s.EncryptionProviderConfigFilepath,
+						s.dynamicTransformers,
+						c.DrainedNotify(),
```
`Run` is invoked in the hook. So it makes sense to propagate hook context inside it.

nilekhc commented:
/cc @liggitt

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -213,12 +218,42 @@ func (s *EtcdOptions) Complete(storageObjectCountTracker flowcontrolrequest.Stor
+					"kms-encryption-config",
+					s.EncryptionProviderConfigFilepath,
+					s.dynamicTransformers,
+					hookContext.StopCh,
```
All `LoadEncryptionConfig` calls should use the same channel.

```suggestion
					stopCh,
```

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -247,16 +282,16 @@ func (s *EtcdOptions) ApplyWithStorageFactoryTo(factory serverstorage.StorageFac
+		factory = &transformerStorageFactory{
+			delegate:            factory,
+			dynamicTransformers: s.dynamicTransformers,
 		}
 	}
 
-	if len(s.transformerOverrides) > 0 {
-		factory = &transformerStorageFactory{
-			delegate:             factory,
-			transformerOverrides: s.transformerOverrides,
+	if !s.SkipHealthEndpoints {
+		if err := s.addEtcdHealthEndpoint(c); err != nil {
+			return err
 		}
 	}
```
Swap the order of these `if` statements back so that the git diff is smaller.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -513,3 +515,135 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+// DynamicTransformers holds transformers that are dynamically updated.
+type DynamicTransformers struct {
+	transformTracker *atomic.Value
+	stopCh           <-chan struct{}
```
Delete this unused field.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -513,3 +515,135 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+	oldTransformTracker := d.transformTracker.Swap(newTransformTracker).(*transformTracker)
+
+	oldTransformTracker.closing.Store(true)
+	if oldTransformTracker.inFlight.Load() == 0 {
+		oldTransformTracker.closeTransformers()
+	}
+}
+
+var _ value.Transformer = &resourceTransformer{}
+
+type resourceTransformer struct {
+	resource         schema.GroupResource
+	transformTracker *atomic.Value
+}
+
+func (r *resourceTransformer) TransformFromStorage(ctx context.Context, data []byte, dataCtx value.Context) ([]byte, bool, error) {
+	// this might cause race condition between two goroutines calling this method at the same time.
+	// due to this old transformer might be closed before they are done serving the request.
+	// ToDo: come up with a better solution to make sure the transformer is not closed when in use.
+	tracker := r.transformTracker.Load().(*transformTracker)
+
+	tracker.inFlight.Add(1)
+	defer func() {
+		if tracker.inFlight.Add(-1) == 0 && tracker.closing.Load() {
+			tracker.closeTransformers()
+		}
+	}()
+
+	if tracker.closing.Load() && tracker.inFlight.Load() == 0 {
+		tracker.closeTransformers()
+		return nil, false, fmt.Errorf("transformer is closing")
+	}
+
+	transformer := tracker.transformerOverrides[r.resource]
+	if transformer == nil {
+		transformer = identity.NewEncryptCheckTransformer()
+	}
+
+	return transformer.TransformFromStorage(ctx, data, dataCtx)
+}
```
@liggitt any ideas on how to make this not racey without adding a lock?  It is trivial to fix with a lock, but I know we would not be okay with a regular mutex here, though maybe a read write mutex would be okay (i.e. `SafeWaitGroup`).  One questionable idea I had was to drop all the in flight tracking and simply sleep for a minute at the end of `.Set` before calling `oldTransformTracker.closeTransformers()`...

nilekhc replied:
Hi @liggitt, would love to hear your thoughts on this. Thanks.

nilekhc replied:
updates as per feedback.

nilekhc commented on test/integration/controlplane/transformation/kms_transformation_test.go:
```diff
@@ -277,6 +278,64 @@ resources:
+    - secrets
+    providers:
+    - kms:
+       name: another-kms-provider
```
added configmaps to the hot reload tests.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -513,3 +515,130 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+	}
+
+	// send swap complete signal
+	d.SwapNotification <- true
```
@enj Is there a better way to do this? I tried returning a function with `<-chan struct{}`, but that was getting convoluted as we need to do this for every `.Set` call.

nilekhc replied:
Removed and implemented with another approach

nilekhc commented on test/integration/controlplane/transformation/kms_transformation_test.go:
```diff
@@ -279,6 +281,228 @@ resources:
+		t.Fatalf("expected secret to be prefixed with %s, but got %s", wantPrefix, rawEnvelope)
+	}
+
+	rawConfigmapEnvelope, err := test.getRawResourceFromETCD(test.getETCDPathForResource("", "configmaps", testConfigmap, testNamespace))
```
I'll rebase for this if #113015 gets merged before.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -513,3 +515,149 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+
+	oldTransformTracker.closing.Store(true)
+	if oldTransformTracker.inFlight.Load() == 0 {
+		// oldTransformTracker.closeTransformers()
```
@enj This intermittently prevents the API server from starting. It's timing out on health checks and thus causing flake in CI. Any idea?

nilekhc replied:
The flake is processing/swapping transformers even before the API server gets a chance to start. When we swap transformers we also close the old ones which causes healthz to timeout.

nilekhc replied:
updated content hash logic

nilekhc commented:
/cc @ritazh @aramase @liggitt

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -513,3 +515,149 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+// DynamicTransformers holds transformers that are dynamically updated.
+type DynamicTransformers struct {
+	transformTracker *atomic.Value
+	SwapNotification *notificationChannel
```
for integration tests, we can verify the new config is effective either by 
1. polling write / check what hit etcd
2. waiting for the backend referenced by the new config to get called

that would let us drop the SwapNotification from this PR

nilekhc replied:
added polling check.

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,266 @@
+	"sync/atomic"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
```
there are edges around handling of paths containing symlinks we want to make sure we handle (since that's how a mounted-in configmap would surface). I *think* we have to watch parent dirs as well in case they are symlinks

nilekhc replied:
Added tests to handle EC file updated by
- truncate (os.WriteFile)
- deleteAndCreate (os.Remove and then os.Create)
- move (write new config in another tmp dir then os.Rename to move it to dir being watched by watchers)
- symLink (created config in parentDir, created symLink of same in dir being watched by watchers, and then update parent dir.)

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -513,3 +515,149 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+}
+
+// Set sets the transformer overrides. This method is not go routine safe and must only be called by the same, single caller throughout the lifetime of this object.
+func (d *DynamicTransformers) Set(
```
if we plumb the max timeout of the set of transformers here, I think we can just wait a multiple of that timeout before closing the replaced transformers and simplify to drop the tracking

```go
Set(newTransformers, newTimeout) {
    oldTransformers, oldTimeout := swap(new, newTimeout)
    go func(){
        // give time for anything with a handle to the old transformers to begin an in-flight and complete with the timeout
        time.Sleep(... 2 * oldTimeout ...)
        // in-flights would now be complete or failed with timeout
        oldTransformers.close()
    }()
}
```

nilekhc replied:
implemented.

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,266 @@
+		closeTransformers()
+		return err
+	}
+
```
should we also health check the new transformers before making them effective?

nilekhc replied:
added health check before calling `Set` to make sure new transformers are healthy.

nilekhc commented on test/integration/controlplane/transformation/kms_transformation_test.go:
```diff
@@ -277,6 +278,64 @@ resources:
+    providers:
+    - kms:
+       name: another-kms-provider
+       cachesize: 1000
```
Created https://github.com/kubernetes/kubernetes/issues/113497 for tracking.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -94,13 +96,16 @@ func (p *kmsv2PluginProbe) toHealthzCheck(idx int) healthz.HealthChecker {
 
-func LoadEncryptionConfig(filepath string, stopCh <-chan struct{}) (map[schema.GroupResource]value.Transformer, []healthz.HealthChecker, error) {
-	config, err := loadConfig(filepath)
+// LoadEncryptionConfig parses encryption configuration file and returns prefix transformers and health checks.
```
Comment is out of date now.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -250,14 +255,26 @@ func loadConfig(filepath string) (*apiserverconfig.EncryptionConfiguration, erro
 	}
 
-	return config, validation.ValidateEncryptionConfiguration(config).ToAggregate()
+	// compare KMS timeout and determine max timeout
+	var maxTimeout time.Duration
+	for _, resourceConfig := range config.Resources {
+		for _, providerConfig := range resourceConfig.Providers {
+			if providerConfig.KMS != nil {
+				if providerConfig.KMS.Timeout.Duration > maxTimeout {
+					maxTimeout = providerConfig.KMS.Timeout.Duration
+				}
+			}
+		}
+	}
```
Move this to

https://github.com/kubernetes/kubernetes/blob/2ab2698e4c33aa0c103ebd0459ccfbef486b7e3c/staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go#L285-L289

It is more code to funnel the information up but I prefer all the KMS logic in one place.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -250,14 +255,26 @@ func loadConfig(filepath string) (*apiserverconfig.EncryptionConfiguration, erro
+		}
+	}
+
+	return config, fmt.Sprintf("%x", sha256.Sum256(data)), maxTimeout, validation.ValidateEncryptionConfiguration(config).ToAggregate()
```
Create a shared function for the hash logic `fmt.Sprintf("%x", sha256.Sum256(data))`

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -513,3 +530,91 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+	closeTransformers    context.CancelFunc
+}
+
+type notificationChannel struct {
+	once sync.Once
+	ch   chan struct{}
+}
```
Remove.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -513,3 +530,91 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+	ch   chan struct{}
+}
+
+// NewDynamicTransformers initializes DynamicTransformers when api-server starts.
```
Remove this comment, it doesn't contain meaningful information.

nilekhc replied:
updated to keep linters happy

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -513,3 +530,91 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+}
+
+// NewDynamicTransformers initializes DynamicTransformers when api-server starts.
+func NewDynamicTransformers(
```
This should take `maxTimeout` as input.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -513,3 +530,91 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+	// this ensures that we wait for any in-flight requests to finish
+	// by the time we spawn this go routine, the new transformers have already been set and will be used for new requests
+	go func() {
+		time.Sleep(maxTimeout * 2)
```
This should come from the old config:

```suggestion
		time.Sleep(oldTransformTracker.maxTimeout * 2)
```

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -513,3 +530,91 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+	transformTracker *atomic.Value
+}
+
+type transformTracker struct {
```
This should track the timeout.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -250,14 +255,26 @@ func loadConfig(filepath string) (*apiserverconfig.EncryptionConfiguration, erro
 	}
 
-	return config, validation.ValidateEncryptionConfiguration(config).ToAggregate()
+	// compare KMS timeout and determine max timeout
+	var maxTimeout time.Duration
```
Add unit tests that cover the `0`, default, less than default, greater than default, etc cases.

nilekhc replied:
added tests

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -62,9 +63,11 @@ type EtcdOptions struct {
-	kmsPluginHealthzChecks []healthz.HealthChecker
+	complete                                    bool
+	dynamicTransformers                         *encryptionconfig.DynamicTransformers
+	KMSTransformersSwapNotificationRegistration func()
+	KMSTransformersSwapNotified                 func() <-chan struct{}
```
Remove.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -213,12 +220,46 @@ func (s *EtcdOptions) Complete(storageObjectCountTracker flowcontrolrequest.Stor
 	if len(s.EncryptionProviderConfigFilepath) != 0 {
-		transformerOverrides, kmsPluginHealthzChecks, err := encryptionconfig.LoadEncryptionConfig(s.EncryptionProviderConfigFilepath, stopCh)
+		ctx, closeTransformers := wait.ContextForChannel(stopCh)
+		transformerOverrides, kmsPluginHealthzChecks, configContentHash, _, err := encryptionconfig.LoadEncryptionConfig(s.EncryptionProviderConfigFilepath, ctx.Done())
```
Timeout needs to be passed to `NewDynamicTransformers`

```suggestion
		transformerOverrides, kmsPluginHealthzChecks, configContentHash, timeout, err := encryptionconfig.LoadEncryptionConfig(s.EncryptionProviderConfigFilepath, ctx.Done())
```

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,273 @@
+	dynamicTransformers *encryptionconfig.DynamicTransformers,
+	configContentHash string,
+	stopCh <-chan struct{},
+) (*DynamicKMSEncryptionConfigContent, error) {
```
This function cannot error anymore.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,273 @@
+	return true
+}
+
+func (d *DynamicKMSEncryptionConfigContent) computeHash(data []byte) string {
+	return fmt.Sprintf("%x", sha256.Sum256(data))
+}
```
Same comment about shared function.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,273 @@
+// workqueueKey is the dummy key used to process change in encryption config file.
+const workqueueKey = "key"
+
+// Runner provides an interface to start encryption config hot reload controller.
+type Runner interface {
+	// Run runs the controller.
+	Run(ctx context.Context)
+}
```
Is this used?

nilekhc replied:
Nope. I was planning to use it with different methods but we pivoted a lot and is not required anymore. Removed.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,273 @@
+		return nil
+	}
+
+	err = d.processEncryptionConfig()
+	if err != nil {
+		return err
+	}
+
+	// update local copy of recent config content once processing is successful.
+	d.encryptionConfig.Store(encryptionConfigHash)
```
Hash needs to come from `encryptionconfig.LoadEncryptionConfig`

```suggestion
	actualEncryptionConfigHash, err := d.processEncryptionConfig()
	if err != nil {
		return err
	}

	// update local copy of recent config content once processing is successful.
	d.encryptionConfig.Store(actualEncryptionConfigHash)
```

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,273 @@
+	if len(healthChecks) > 0 {
+		// test with only first health check. We will be moving towards a single health check for all kms providers.
+		healthCheck := healthChecks[0]
+		healthCheckResult := healthCheck.Check(nil)
```
This needs a valid request otherwise it will panic with KMS v2.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,273 @@
+	return nil
+}
+
+func (d *DynamicKMSEncryptionConfigContent) validateNewTransformersHealth(healthChecks []healthz.HealthChecker) error {
```
This needs to enforce a timeout, say `5 * time.Minute`.  Ideally this would be canceled early if the file hash changes but it should give at least `2 * time.Minute` to allow the check to complete in case the file is being mutated "often."

enj reviewed (commented):
Still need to look at the tests more closely.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/util/util.go:
```diff
@@ -52,3 +53,8 @@ func ParseEndpoint(endpoint string) (string, error) {
 	return u.Path, nil
 }
+
+// ComputeHash returns the SHA256 hash of the given data.
```
@enj lmk if there is a better place for this shared function.

aramase replied:
We can add a small unit test for this. Looking at some of the other packages:

```go
func TestHash(t *testing.T) {
	// hash the empty string to be sure that sha256 is being used
	expect := "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
	sum := ComputeHash("")
	if expect != sum {
		t.Errorf("expected hash %q but got %q", expect, sum)
	}
}
```

nilekhc replied:
added

nilekhc replied:
> We can add a small unit test for this. Looking at some of the other packages:
> 
> ```go
> func TestHash(t *testing.T) {
> 	// hash the empty string to be sure that sha256 is being used
> 	expect := "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
> 	sum := ComputeHash("")
> 	if expect != sum {
> 		t.Errorf("expected hash %q but got %q", expect, sum)
> 	}
> }
> ```

added

nilekhc replied:
It came up in discussing the need for a hash with @enj. There can be keys directly embedded in the config file. So I am thinking of keeping it to sha.

I moved this to the encryptionconfig package where we load config.

ritazh commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -94,45 +96,54 @@ func (p *kmsv2PluginProbe) toHealthzCheck(idx int) healthz.HealthChecker {
-			return nil, nil, err
+			return nil, nil, timeout, err
 		}
+		kmsMaxTimeout = timeout
```
Where do you compare the values to get the max timeout? The current implementation will always return the last timeout as max.

nilekhc replied:
I compare it here: https://github.com/kubernetes/kubernetes/pull/112050/files#diff-8ca5bf5da3d263c7a4d934647ad0f3d929256c9b399d305576b0fc84c6b966deR303-R306

Unit tests seem to agree with this:
https://github.com/kubernetes/kubernetes/pull/112050/files#diff-f2f067d6aa03c572d14285964d389066c885f5f54df7dc7d779ccecdc50f7b75R259

ritazh replied:
prefixTransformersAndProbes returns the max timeout from one resourceConfig's providers. where do you compare the max timeout for all resourceConfigs?

nilekhc replied:
Ah, good catch. Updated the logic and tests. Thanks, @ritazh!

aramase commented on test/integration/controlplane/transformation/kms_transformation_test.go:
```diff
@@ -279,6 +281,470 @@ resources:
 
+// TestECHotReload is an integration test that verifies hot reload of KMS encryption config works.
+// This test asserts following scenarios:
+// 1. start at KMS 1
```
Instead of `KMS 1` and `KMS 2`, let's have the actual kms provider names in the comments

aramase commented on test/integration/controlplane/transformation/kms_transformation_test.go:
```diff
@@ -279,6 +281,470 @@ resources:
+       cachesize: 1000
+       endpoint: unix:///@kms-provider.sock
+`
+	// providerName := "kms-provider"
```
```suggestion
```

aramase commented on test/integration/controlplane/transformation/kms_transformation_test.go:
```diff
@@ -279,6 +281,470 @@ resources:
+		t.Fatalf("Failed to create test secret, error: %v", err)
+	}
+
+	// create a new secret in default namespace. This is to asset cluster wide read works after hot reload.
```
```suggestion
	// create a new secret in default namespace. This is to assert cluster wide read works after hot reload.
```

aramase commented on test/integration/controlplane/transformation/kms_transformation_test.go:
```diff
@@ -279,6 +281,470 @@ resources:
+	// create a new secret in default namespace. This is to asset cluster wide read works after hot reload.
+	_, err = test.createSecret(fmt.Sprintf("%s-%s", testSecret, "1"), "default")
+	if err != nil {
+		t.Fatalf("Failed to create test secret, error: %v", err)
```
```suggestion
		t.Fatalf("Failed to create test secret in default namespace, error: %v", err)
```

aramase commented on test/integration/controlplane/transformation/kms_transformation_test.go:
```diff
@@ -279,6 +281,470 @@ resources:
+    - identity: {}
+`
+	// start new KMS Plugin
+	// newProviderName := "new-kms-provider"
```
```suggestion
```

aramase commented on test/integration/controlplane/transformation/kms_transformation_test.go:
```diff
@@ -279,6 +281,470 @@ resources:
+
+	wantPrefix := "k8s:enc:kms:v1:new-kms-provider:"
+
+	// implementing this brut force approach instead of fancy channel notification to avoid test specific code in prod.
```
```suggestion
	// implementing this brute force approach instead of fancy channel notification to avoid test specific code in prod.
```

aramase commented on test/integration/controlplane/transformation/kms_transformation_test.go:
```diff
@@ -279,6 +281,470 @@ resources:
+		}
+	}
+
+	//get configmaps
```
```suggestion
	// get configmaps
```

aramase commented on test/integration/controlplane/transformation/kms_transformation_test.go:
```diff
@@ -279,6 +281,470 @@ resources:
+		t.Fatalf("failed to read %s from etcd: %v", test.getETCDPathForResource(test.storageConfig.Prefix, "", "configmaps", testConfigmap, testNamespace), err)
+	}
+
+	// assert configmap
```
```suggestion
	// assert prefix for configmap
```

aramase commented on test/integration/controlplane/transformation/kms_transformation_test.go:
```diff
@@ -279,6 +281,470 @@ resources:
+       cachesize: 1000
+       endpoint: unix:///@kms-provider.sock
+`
+			// providerName := "kms-provider"
```
```suggestion
```

aramase commented on test/integration/controlplane/transformation/kms_transformation_test.go:
```diff
@@ -279,6 +281,470 @@ resources:
+    - identity: {}
+`
+			// start new KMS Plugin
+			// newProviderName := "new-kms-provider"
```
```suggestion
```

aramase commented on test/integration/controlplane/transformation/kms_transformation_test.go:
```diff
@@ -279,6 +281,470 @@ resources:
+				if err := os.Remove(path.Join(test.configDir, encryptionConfigFileName)); err != nil {
+					t.Fatalf("failed to remove encryption config, err: %v", err)
+				}
+				// os.WriteFile creates a new file if it doesn't exist already
```
The test is using `os.Create`, so this comment is no longer required?

nilekhc replied:
removed

aramase commented on test/integration/controlplane/transformation/kms_transformation_test.go:
```diff
@@ -279,6 +281,470 @@ resources:
+
+			wantPrefix := "k8s:enc:kms:v1:new-kms-provider:"
+
+			// implementing this brut force approach instead of fancy channel notification to avoid test specific code in prod.
```
```suggestion
			// implementing this brute force approach instead of fancy channel notification to avoid test specific code in prod.
```

aramase commented on test/integration/controlplane/transformation/kms_transformation_test.go:
```diff
@@ -279,6 +281,470 @@ resources:
+			}
+
+			swapErr = fmt.Errorf("expected secret to be prefixed with %s, but got %s", wantPrefix, rawEnvelope.Kvs[0].Value)
+			return false, swapErr
```
IIRC, returning an error would stop the polling right? Don't we want to poll until timeout to check if it's swapped?

nilekhc replied:
Good catch! Updated the code.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -513,3 +530,90 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+	maxTimeout           time.Duration
+}
+
+// NewDynamicTransformers ...
```
I have been asked to remove this, but then kept `...` to make linters happy. 😅

https://github.com/kubernetes/kubernetes/pull/112050#discussion_r1010639515

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go:
```diff
@@ -256,6 +256,155 @@ func TestEncryptionProviderConfigCorrect(t *testing.T) {
+func TestKMSTimeout(t *testing.T) {
+	testCases := []struct {
+		name            string
+		expectError     bool
```
Actually, we already check it in `TestKMSPluginHealthz`. I could just remove this. The error string here is no different here. 

This is mainly to test the value of the timeout being returned.

nilekhc replied:
Removed error validation since we are already doing it in other tests

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,273 @@
+	// ignoring content hash as it's already checked in loadEncryptionConfig.
+	transformerOverrides, healthChecks, contentHash, maxTimeout, err := encryptionconfig.LoadEncryptionConfig(d.filePath, ctx.Done())
+	if err != nil {
+		klog.Warning("KMS encryption configuration is not valid.", err)
```
I am closing the transformer, so logging as well to help logs/debugging easier in case of error.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,273 @@
+	// test if new transformers has good health.
+	err = d.validateNewTransformersHealth(healthChecks)
+	if err != nil {
+		klog.Warning("health check for new transformers failed.", err)
```
Updated to `Warningf`

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,272 @@
+		stopCh:              stopCh,
+		queue:               workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), fmt.Sprintf("%s-hot-reload", name)),
+	}
+	encryptionConfig.encryptionConfig.Store(configContentHash)
```
I think you still need to add to the queue to not miss any events between the initial `Load` and when `Run` is called.

```suggestion
	encryptionConfig.encryptionConfig.Store(configContentHash)
	encryptionConfig.queue.Add(workqueueKey)
```

nilekhc replied:
added

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -281,7 +318,9 @@ func (s *EtcdOptions) addEtcdHealthEndpoint(c *server.Config) error {
 	}))
 
-	c.AddHealthChecks(s.kmsPluginHealthzChecks...)
+	if s.kmsPluginHealthzCheck != nil {
+		c.AddHealthChecks(s.kmsPluginHealthzCheck)
+	}
```
```suggestion
	if s.dynamicTransformers != nil {
		c.AddHealthChecks(s.dynamicTransformers)
	}
```

nilekhc replied:
updated

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -513,3 +561,117 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+}
+
+// Name implements healthz.HealthChecker
+func (d *DynamicTransformers) Name() string {
+	return d.transformTracker.Load().(*transformTracker).kmsPluginHealthzCheck.Name()
+}
```
This gives the false impression that this value can change over time, but it cannot because the API server wiring captures it on startup.  Instead we should:

```suggestion
func (d *DynamicTransformers) Name() string {
	return d.name
}
```

and in `NewDynamicTransformers` set `dynamicTransformers.name` to `kmsPluginHealthzCheck.Name()`

nilekhc replied:
done

nilekhc commented:
Hi @liggitt, Could you please review the overall PR when you get a chance? Thanks!

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -63,9 +63,9 @@ type EtcdOptions struct {
 	WatchCacheSizes []string
 
 	// complete guards fields that must be initialized via Complete before the Apply methods can be used.
```
shouldn't this commit be dropped if this is based on the latest version of https://github.com/kubernetes/kubernetes/pull/113529?

nilekhc replied:
I think we still need the `complete` bool to make sure we call the `Complete()` method exactly once, don't we?

liggitt replied:
that's not part of that commit, is it? (I left this comment on the first line of that commit... I'm not suggesting the `complete` field be dropped)

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,299 @@
+	// This should be executed after restarting the watch (if applicable) to ensure no file event will be missing.
+	defer d.queue.Add(workqueueKey)
+
+	// return if file has not been removed or renamed.
+	if event.Op&(fsnotify.Remove|fsnotify.Rename) == 0 {
+		return nil
+	}
+
+	if err := watcher.Remove(d.filePath); err != nil {
+		klog.V(2).InfoS("Failed to remove file watch, it may have been deleted", "file", d.filePath, "err", err)
+	}
+	if err := watcher.Add(d.filePath); err != nil {
+		return fmt.Errorf("error adding watch for file %s: %v", d.filePath, err)
+	}
```
I don't know enough to know if this covers all cases where we'd need to re-add the file watch... wondering who would know

nilekhc replied:
The basic idea behind this is to cover when somebody instead of just updating an existing file, deletes the old one and adds a new file (with updated content and the same name).

Though additional thoughts are always welcome. 🙂 cc @enj @ritazh @aramase

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,299 @@
+		return true
+	}
+
+	err = d.updateTransformers(
```
the nesting of logic and calling closeTransformers() at uneven levels in error cases (processNextWorkItem → updateTransformers → healthz) is pretty hard to follow

it would be easier to be confident we've handled unchanged and error cases if this was called in a single place in processNextWorkItem, and processNextWorkItem was responsible for the checks and cleanup in error cases

```go
ctx, closeTransformers := wait.ContextForChannel(d.stopCh)
var (
  updatedEffectiveConfig bool
  err error
)
defer func(){
  if updatedEffectiveConfig {
    // TODO: increment success metric
  } else {
    // avoid leaking on error or unchanged
    closeTransformers()
  }

  if err != nil {
    // TODO: increment failure metric
  }
}()

..., configChanged, err = processEncryptionConfig(...)
if err != nil {
  return
}
if !configChanged {
  return
}
err = healthz(...)
if err != nil {
  return
}
d.dynamicTransformers.Set(...)
d.encryptionConfigHash = processedConfigHash
updatedEffectiveConfig = true
return
```

nilekhc replied:
I had it consolidated before but then separated it to make healthz testable. But this is better. I updated it. Thanks!

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -82,57 +84,99 @@ type kmsv2PluginProbe struct {
+		kmsv2 = kmsv2 || kmsv2Used
+		kmsv1 = kmsv1 || kmsv1Used
+
+		resourceProviderTimeouts = append(resourceProviderTimeouts, timeout)
```
Done

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -213,12 +227,48 @@ func (s *EtcdOptions) Complete(storageObjectCountTracker flowcontrolrequest.Stor
+			func(hookContext server.PostStartHookContext) error {
+				// get context
+				// we ignore the cancel func because this context should only be canceled when stopCh is closed
+				ctx, _ := wait.ContextForChannel(hookContext.StopCh)
```
Since the hot reload controller is triggered from the postStartHook I have propagated that context to tie the controller lifetime to the hook. For others like LoadEncryptionConfig (which also creates new gRPC services), and initialization of the dynamicTransformers I have kept the server lifetime context.

nilekhc replied:
We simplify this by passing the api-server stopCh to all the hot reload code paths to make sure they all are tied to the api-server lifecycle.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -213,12 +227,48 @@ func (s *EtcdOptions) Complete(storageObjectCountTracker flowcontrolrequest.Stor
+		s.dynamicTransformers = encryptionconfig.NewDynamicTransformers(transformerOverrides, kmsPluginHealthzChecks, closeTransformers, kmsMaxTimeout, totalKMSPlugins)
+
+		if !s.EncryptionProviderConfigAutomaticReload {
+			s.kmsPluginHealthzChecks = kmsPluginHealthzChecks
```
After rebasing on https://github.com/kubernetes/kubernetes/pull/113529 while running unit tests I could see separate health check names viz `kms-provider-0`, `kms-provider-1`. But when reload is set to true we get `kms-providers`

When reload is not true and if we pass the dynamic health check then health checker() fails as dynamicTransformers could not distinguish between idx `0`, `1` etc.

nilekhc replied:
That's exactly what I have done. 🙂

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -213,12 +227,48 @@ func (s *EtcdOptions) Complete(storageObjectCountTracker flowcontrolrequest.Stor
+
+		// add post start hook to start hot reload controller
+		// adding this hook here will ensure that it gets configured exactly once
+		err = addPostStartHook(
```
makes sense. Done.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -213,12 +227,48 @@ func (s *EtcdOptions) Complete(storageObjectCountTracker flowcontrolrequest.Stor
+		)
+		if err != nil {
+			klog.Warningf("failed to add post start hook for kms encryption config hot reload controller: %v", err)
 			return err
```
Updated to be `Errorf`.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -281,7 +331,13 @@ func (s *EtcdOptions) addEtcdHealthEndpoint(c *server.Config) error {
 	}))
 
-	c.AddHealthChecks(s.kmsPluginHealthzChecks...)
+	if s.EncryptionProviderConfigAutomaticReload {
+		if s.dynamicTransformers != nil {
+			c.AddHealthChecks(s.dynamicTransformers)
+		}
+	} else {
+		c.AddHealthChecks(s.kmsPluginHealthzChecks...)
+	}
```
Same as above. https://github.com/kubernetes/kubernetes/pull/112050#discussion_r1014687256

Will check with @enj again.

nilekhc replied:
updated per - https://github.com/kubernetes/kubernetes/pull/112050#discussion_r1015849861

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,299 @@
+	}
+	defer d.queue.Done(key)
+
+	// get context to close the old transformers.
```
That's right. Updated the comment.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,299 @@
+		// close transformers to avoid leaking resources.
+		closeTransformers()
+
+		utilruntime.HandleError(fmt.Errorf("error processing encryption config file %s: %v", d.filePath, err))
```
added a note.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,299 @@
+	filePath string
+
+	// encryptionProviderConfigAutomaticReload is the api-server flag to enable automatic reload of the encryption config file.
+	encryptionProviderConfigAutomaticReload bool
```
I was hoping to propagate the api-server flag as is to the `LoadEncryptionConfig` method in this controller. But I guess I can explicitly pass `true` since this code will only execute if `reload=true`.

https://github.com/kubernetes/kubernetes/pull/112050/files#diff-b41a7383f2bf663071c77c39cc1e4117289399a3e25a3ec59e791b5029d9ff72R234

nilekhc replied:
removed.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,299 @@
+	encryptionProviderConfigAutomaticReload bool
+
+	// encryptionConfig stores last successfully read encryption config file content.
+	encryptionConfig atomic.Value
```
makes sense. updated.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,299 @@
+	// dynamicTransformers updates the transformers when encryption config file changes.
+	dynamicTransformers *encryptionconfig.DynamicTransformers
+
+	stopCh <-chan struct{}
```
yup. added a comment.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,299 @@
+		d.queue.Forget(key)
+		return true
+	}
+
```
added a note.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,299 @@
+	}
+
+	// if transformerOverrides is nil, it means that the encryption config file has not changed.
+	if transformerOverrides == nil {
```
updated

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -152,7 +196,14 @@ func getTransformerOverridesAndKMSPluginProbes(config *apiserverconfig.Encryptio
+	// calculate the max timeout among all transformers
+	for _, timeout := range resourceProviderTimeouts {
+		if timeout > kmsMaxTimeout {
+			kmsMaxTimeout = timeout
```
updated the logic.

enj reviewed (commented):
Will provide more feedback soon.

nilekhc committed:
feat: implements kms encryption config hot reload

This change enables hot reload of encryption config file when api server
flag --encryption-provider-config-automatic-reload is set to true. This
allows the user to change the encryption config file without restarting
kube-apiserver. The change is detected by polling the file and is done
by using fsnotify watcher. When file is updated it's process to generate
new set of transformers and close the old ones.

Signed-off-by: Nilekh Chaudhari <1626598+nilekhc@users.noreply.github.com>

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -82,57 +84,99 @@ type kmsv2PluginProbe struct {
+// It may launch multiple go routines whose lifecycle is controlled by stopCh.
+// If reload is true, or KMS v2 plugins are used with no KMS v1 plugins, the returned slice of health checkers will always be of length 1.
+// LoadEncryptionConfig parses encryption configuration file and returns prefix transformers and health checks.
+func LoadEncryptionConfig(filepath string, reload bool, stopCh <-chan struct{}) (map[schema.GroupResource]value.Transformer, []healthz.HealthChecker, string, time.Duration, int, error) {
```
Added struct.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -213,12 +227,50 @@ func (s *EtcdOptions) Complete(storageObjectCountTracker flowcontrolrequest.Stor
+			// add post start hook to start hot reload controller
+			// adding this hook here will ensure that it gets configured exactly once
+			err = addPostStartHook(
+				"start-kms-encryption-config-hot-reloader",
```
Done

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -281,7 +333,13 @@ func (s *EtcdOptions) addEtcdHealthEndpoint(c *server.Config) error {
 	}))
 
-	c.AddHealthChecks(s.kmsPluginHealthzChecks...)
+	if s.EncryptionProviderConfigAutomaticReload {
+		if s.dynamicTransformers != nil {
+			c.AddHealthChecks(s.dynamicTransformers)
+		}
+	} else {
```
makes sense. updated.

aramase commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -113,60 +115,88 @@ func (h *kmsv2PluginProbe) toHealthzCheck(idx int) healthz.HealthChecker {
+	config, contentHash, err := loadConfig(filepath)
 	if err != nil {
-		return nil, nil, fmt.Errorf("error while parsing file: %w", err)
+		return nil, fmt.Errorf("error while parsing file: %v", err)
```
```suggestion
		return nil, fmt.Errorf("error while parsing file: %w", err)
```

aramase commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -113,60 +115,88 @@ func (h *kmsv2PluginProbe) toHealthzCheck(idx int) healthz.HealthChecker {
 	}
 
-	transformers, kmsHealthChecks, kmsUsed, err := getTransformerOverridesAndKMSPluginHealthzCheckers(config, stopCh)
+	transformers, kmsHealthChecks, kmsv2, kmsv1, kmsMaxTimeout, err := getTransformerOverridesAndKMSPluginHealthzCheckers(config, stopCh)
```
Is there a reason we are changing from the struct to individual bool vars? IMO the struct is better and we don't have to see the function to know what are the vars returned.

nilekhc replied:
@enj  any thoughts on this change?

enj replied:
Please revert back to the use of the `kmsState` struct in the inner functions.

enj replied:
And funnel the max timeout up by adding a new field for it in the `kmsState` struct.

aramase commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -113,60 +115,88 @@ func (h *kmsv2PluginProbe) toHealthzCheck(idx int) healthz.HealthChecker {
-	var kmsUsed kmsState
+	var kmsv2, kmsv1 bool
+	var kmsMaxTimeout time.Duration
+	// var totalKMSPlugins int
```
```suggestion
```

aramase commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -113,60 +115,88 @@ func (h *kmsv2PluginProbe) toHealthzCheck(idx int) healthz.HealthChecker {
 		}
-		kmsUsed.v1Used = kmsUsed.v1Used || used.v1Used
-		kmsUsed.v2Used = kmsUsed.v2Used || used.v2Used
+		// totalKMSPlugins += kmsPlugins
```
```suggestion
```

aramase commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -262,19 +292,19 @@ func isKMSv2ProviderHealthy(name string, response *envelopekmsv2.StatusResponse)
 	f, err := os.Open(filepath)
 	if err != nil {
-		return nil, fmt.Errorf("error opening encryption provider configuration file %q: %w", filepath, err)
+		return nil, "", fmt.Errorf("error opening encryption provider configuration file %q: %v", filepath, err)
```
```suggestion
		return nil, "", fmt.Errorf("error opening encryption provider configuration file %q: %w", filepath, err)
```

aramase commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -262,19 +292,19 @@ func isKMSv2ProviderHealthy(name string, response *envelopekmsv2.StatusResponse)
 	data, err := io.ReadAll(f)
 	if err != nil {
-		return nil, fmt.Errorf("could not read contents: %w", err)
+		return nil, "", fmt.Errorf("could not read contents: %v", err)
```
```suggestion
		return nil, "", fmt.Errorf("could not read contents: %w", err)
```

aramase commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -319,31 +351,43 @@ func prefixTransformersAndProbes(config apiserverconfig.ResourceConfiguration, s
 
+			// request might take max timeout to work through all the KMS plugins
+			// sum the timeouts for all KMS plugins to ensure that we are allowing sufficient time before closing old transformers
+			kmsMaxTimeoutForResource += provider.KMS.Timeout.Duration
```
Shouldn't the max timeout be sufficient to accommodate timeout for plugins? Why are we adding here?

nilekhc replied:
https://github.com/kubernetes/kubernetes/pull/112050#discussion_r1014649817

nilekhc replied:
updated as per - https://github.com/kubernetes/kubernetes/pull/112050#discussion_r1016056598

aramase commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -319,31 +351,43 @@ func prefixTransformersAndProbes(config apiserverconfig.ResourceConfiguration, s
+			// sum the timeouts for all KMS plugins to ensure that we are allowing sufficient time before closing old transformers
+			kmsMaxTimeoutForResource += provider.KMS.Timeout.Duration
+
+			// // compare KMS timeout and determine max timeout
+			// if provider.KMS.Timeout.Duration > kmsMaxTimeout {
+			// 	kmsMaxTimeout = provider.KMS.Timeout.Duration
+			// }
+
+			// // increment totalKMSPlugins
+			// totalKMSPlugins++
```
remove these if not used

aramase commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,274 @@
+
+// NewDynamicKMSEncryptionConfiguration returns controller that dynamically reacts to changes in encryption config file.
+func NewDynamicKMSEncryptionConfiguration(
+	name string,
+	filePath string,
```
```suggestion
	name, filePath string,
```

aramase commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,274 @@
+}
+
+func (d *DynamicKMSEncryptionConfigContent) watchEncryptionConfigFile(stopCh <-chan struct{}) error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("error creating fsnotify watcher: %v", err)
+	}
+	defer watcher.Close()
+
+	if err = watcher.Add(d.filePath); err != nil {
+		return fmt.Errorf("error adding watch for file %s: %v", d.filePath, err)
+	}
+
+	for {
+		select {
+		case event := <-watcher.Events:
+			if err := d.handleWatchEvent(event, watcher); err != nil {
+				return err
+			}
+		case err := <-watcher.Errors:
+			return fmt.Errorf("received fsnotify error: %v", err)
+		case <-stopCh:
+			return nil
+		}
```
wrap the errors here with `%w` instead of `%v`?

aramase commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,274 @@
+	ctx, closeTransformers := wait.ContextForChannel(d.stopCh)
+
+	defer func() {
+		if updatedEffectiveConfig {
+			// TODO: increment success metric
+		} else {
+			// avoid leaking if we're not using the newly constructed transformers (due to an error or them not being changed)
+			closeTransformers()
+		}
```
Maybe just handle `if !updatedEffectiveConfig` now and keep the todo instead of keeping the if loop with a TODO?

aramase commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,274 @@
+
+// loadEncryptionConfig processes the next set of content from the file.
+func (d *DynamicKMSEncryptionConfigContent) processEncryptionConfig(ctx context.Context) (
+	encryptionConfiguration *encryptionconfig.EncryptionConfiguration,
+	configChanged bool,
+	err error,
```
Any reason for having named return parameters just for this function? `configChanged` is really just used in the return statements and never set anywhere in code.

nilekhc replied:
To give readability to the code. Returning simply bool is confusing as to what it is used for.

aramase commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,274 @@
+
+	for _, kmsPluginHealthzCheck := range kmsPluginHealthzChecks {
+		// create http get request to health check endpoint.
+		req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("/healthz/%v", kmsPluginHealthzCheck.Name()), nil)
```
```suggestion
		req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("/healthz/%s", kmsPluginHealthzCheck.Name()), nil)
```

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -283,29 +332,43 @@ func prefixTransformersAndProbes(config apiserverconfig.ResourceConfiguration, s
 
+			// request might take max timeout to work through all the KMS plugins
+			// sum the timeouts for all KMS plugins to ensure that we are allowing sufficient time before closing old transformers
+			kmsMaxTimeoutForResource += provider.KMS.Timeout.Duration
```
https://github.com/kubernetes/kubernetes/pull/112050#discussion_r1014649817

enj replied:
@liggitt this seems a bit too clever ...

We already know how many KMS plugins we have via `len(kmsHealthChecks)` so we don't actually need to funnel the count up.  I feel like the math I described in https://github.com/kubernetes/kubernetes/pull/113529#discussion_r1015647113 is way more straightforward.

nilekhc replied:
updated the math as per this.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,274 @@
+			closeTransformers()
+		}
+		if err != nil {
+			// TODO: increment failure metric
```
https://github.com/kubernetes/kubernetes/issues/113704

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,274 @@
+	}
+
+	// update transformers.
+	d.dynamicTransformers.Set(transformerOverrides, closeTransformers, kmsPluginHealthzChecks, kmsMaxTimeout)
```
In `Set` we always pass the new healthz check and `LoadEcncryptionConfig` always returns multiple healthz (if there are various kms plugins).  Why do we want to check the old health check here?

Am I missing anything here?

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,274 @@
+) {
+	encryptionConfig, err := d.readEncryptionFileContent()
+	if err != nil {
+		return nil, nil, "", 0, configChanged, err
```
isn't default false?

enj replied:
Setting a non-zero return value makes me think I need to figure out what that value is on return instead of ignoring it and just looking at the error.  Just like we explicitly pass in `true` for reload, this should explicitly return `false`.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,274 @@
+		return nil, nil, "", 0, configChanged, err
+	}
+	if len(encryptionConfig) == 0 {
+		return nil, nil, "", 0, configChanged, fmt.Errorf("encryption config file is empty %q", d.filePath)
```
same as above

nilekhc replied:
updated

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -284,28 +312,30 @@ func loadConfig(filepath string, reload bool) (*apiserverconfig.EncryptionConfig
-	var kmsUsed kmsState
+	var kmsv2, kmsv1 bool
+	var kmsMaxTimeoutForResource time.Duration
+	// var totalKMSPlugins int
```
Still need to remove

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -555,3 +585,118 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+	return fmt.Sprintf("%x", sha256.Sum256(data))
+}
+
+// Dynamic Transformers
```
```suggestion
```

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -555,3 +585,118 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+	kmsMaxTimeout         time.Duration
+}
+
+// NewDynamicTransformers ...
```
Describe what this struct does.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -113,60 +115,86 @@ func (h *kmsv2PluginProbe) toHealthzCheck(idx int) healthz.HealthChecker {
 	})
 }
 
+// EncryptionConfiguration represents the encryption configuration for the apiserver.
```
```suggestion
// EncryptionConfiguration represents the parsed and normalized encryption configuration for the apiserver.
```

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -223,12 +228,55 @@ func (s *EtcdOptions) Complete(storageObjectCountTracker flowcontrolrequest.Stor
+
+		s.dynamicTransformers = encryptionconfig.NewDynamicTransformers(encryptionConfiguration.Transformers, kmsPluginHealthzCheck, closeTransformers, encryptionConfiguration.KMSMaxTimeout)
+
+		if !s.EncryptionProviderConfigAutomaticReload {
+			s.kmsPluginHealthzChecks = encryptionConfiguration.HealthChecks
+		}
```
```suggestion
		s.kmsPluginHealthzChecks = encryptionConfiguration.HealthChecks
```

nilekhc commented:
/cc @liggitt

enj commented on test/integration/controlplane/transformation/transformation_test.go:
```diff
@@ -250,20 +262,40 @@ func (e *transformTest) getEncryptionOptions(reload bool) []string {
+			return tempDir, parentTempDir, fmt.Errorf("failed to write encryption config file: %v", err)
+		}
+
+		// create symlink
+		if err := os.Symlink(filepath.Join(parentTempDir, encryptionConfigFileName), filepath.Join(tempDir, encryptionConfigFileName)); err != nil {
```
This does not look correct.  I would expect the config directories to be symlinked, not the config files (though we can test both approaches)

```suggestion
		// create a symlink from temp to parent
		if err := os.Symlink(parentTempDir, tempDir); err != nil {
```

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,278 @@
+			return err
+		}
+
+		pollErr := wait.PollImmediate(100*time.Millisecond, 2*time.Minute, func() (bool, error) {
```
2 minutes seems really long... it means if someone updates the config with a bad endpoint, our worker will block for 2 minutes per kms trying to health check that bad endpoint even if they've fixed the config to a good endpoint in the meantime

if we return an error, we'll requeue the config and retry with the latest content, right? trying this for something like ~10 seconds seems better

nilekhc replied:
updated to check for 10s or `kmsPluginCloseGracePeriod` if it's higher.

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,278 @@
+			healthCheckError := kmsPluginHealthzCheck.Check(req)
+			return healthCheckError == nil, nil
+		})
+		if pollErr == wait.ErrWaitTimeout {
```
```suggestion
		if pollErr != nil {
```

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,278 @@
+
+	for _, kmsPluginHealthzCheck := range kmsPluginHealthzChecks {
+		// create http get request to health check endpoint.
+		req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("/healthz/%s", kmsPluginHealthzCheck.Name()), nil)
```
I can't remember... is reusing a request multiple times allowed?

enj reviewed (commented):
I still need to go through these tests but wanted to provide feedback early since we are low on time.

- `staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller_test.go`
- `test/integration/controlplane/transformation/kms_transformation_test.go`

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,278 @@
+	// this code path will only execute if reload=true. So passing true explicitly.
+	encryptionConfiguration, err = encryptionconfig.LoadEncryptionConfig(d.filePath, true, ctx.Done())
+
+	return encryptionConfiguration, true, err
```
there's a time-of-check-time-of-use race here... we're reading the content and computing the hash above, but then LoadEncryptionConfig re-reads the file and recomputes the hash

if a file was written incrementally in place (not a good idea, but possible), our local read/hash could think the file changed, then the LoadEncryptionConfig could see the original content and hash

should we do this instead:
```go
if err != nil {
  return nil, false, err
}
return encryptionConfiguration, d.hasEncryptionConfigChanged(encryptionConfiguration.EncryptionFileContentHash), nil
```

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -223,12 +228,58 @@ func (s *EtcdOptions) Complete(storageObjectCountTracker flowcontrolrequest.Stor
+
+		s.kmsPluginHealthzChecks = encryptionConfiguration.HealthChecks
+
+		var kmsPluginHealthzCheck healthz.HealthChecker
+		if len(encryptionConfiguration.HealthChecks) > 0 {
+			kmsPluginHealthzCheck = encryptionConfiguration.HealthChecks[0]
+		}
```
it's confusing to propagate this into NewDynamicTransformers when it's not going to be used in the non-reload case... I'm trying to think how to make it clearer that there's ~two modes NewDynamicTransformers operates in

* non-reloading: it is not a health checker, Set will never be called, it's only handling delegating transformer lookups for resources. we could construct with `encryptionconfig.NewDynamicTransformers(encryptionConfiguration.Transformers)`
* reloading: it *is* a health checker, Set will be called. we could construct with `encryptionconfig.NewReloadingDynamicTransformers(encryptionConfiguration.Transformers, kmsPluginHealthzCheck, closeTransformers, encryptionConfiguration.KMSTimeout)`

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -223,12 +228,58 @@ func (s *EtcdOptions) Complete(storageObjectCountTracker flowcontrolrequest.Stor
-		s.transformerOverrides = transformerOverrides
-		s.kmsPluginHealthzChecks = kmsPluginHealthzChecks
+
+		s.kmsPluginHealthzChecks = encryptionConfiguration.HealthChecks
```
it's confusing to set this when it's not going to be used in the reloading case

I think we should only have `s.kmsPluginHealthzChecks`, and set it conditional on whether we're reloading or not

```go
var err error
if s.EncryptionProviderConfigAutomaticReload {
  s.dynamicTransformers, err = encryptionconfig.NewReloadingDynamicTransformers(encryptionConfiguration, closeTransformers)
  if err != nil {
    return err
  }
  s.kmsPluginHealthzChecks = []healthz.HealthChecker{s.dynamicTransformers}
  err = addPostStartHook(...
    ...
} else {
  s.dynamicTransformers = encryptionconfig.NewDynamicTransformers(encryptionConfiguration.Transformers)
  s.kmsPluginHealthzChecks = encryptionConfiguration.HealthChecks
}
```

`NewReloadingDynamicTransformers` can ensure there's a single non-nil health check, etc

EtcdOptions#addEtcdHealthEndpoint can stay unchanged and just do `c.AddHealthChecks(s.kmsPluginHealthzChecks...)`

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -223,12 +228,58 @@ func (s *EtcdOptions) Complete(storageObjectCountTracker flowcontrolrequest.Stor
+							s.EncryptionProviderConfigFilepath,
+							s.dynamicTransformers,
+							encryptionConfiguration.EncryptionFileContentHash,
+							stopCh,
```
it's confusing to have stopCh and hookContext.StopCh and ctx (based on hookContext.StopCh), and pass one to NewDynamicKMSEncryptionConfiguration and another to Run...

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -223,12 +228,58 @@ func (s *EtcdOptions) Complete(storageObjectCountTracker flowcontrolrequest.Stor
+		encryptionConfiguration, err := encryptionconfig.LoadEncryptionConfig(s.EncryptionProviderConfigFilepath, s.EncryptionProviderConfigAutomaticReload, ctx.Done())
 		if err != nil {
+			// in case of error, we want to close partially initialized (if any) transformers
+			closeTransformers()
```
are we calling closeTransformers() consistently in the error cases below where we return?

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -223,12 +228,58 @@ func (s *EtcdOptions) Complete(storageObjectCountTracker flowcontrolrequest.Stor
+					return err
+				}
+			} else {
+				klog.Errorf("failed to start kms encryption config hot reload controller. only 1 health check should be available when reload is enabled")
```
shouldn't this return an error rather than log?

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -223,12 +228,58 @@ func (s *EtcdOptions) Complete(storageObjectCountTracker flowcontrolrequest.Stor
+					},
+				)
+				if err != nil {
+					klog.Errorf("failed to add post start hook for kms encryption config hot reload controller: %v", err)
+					return err
```
should this wrap the error rather than log?

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -291,7 +342,11 @@ func (s *EtcdOptions) addEtcdHealthEndpoint(c *server.Config) error {
 	}))
 
-	c.AddHealthChecks(s.kmsPluginHealthzChecks...)
+	if s.EncryptionProviderConfigAutomaticReload {
```
see above comments on making `c.AddHealthChecks(s.kmsPluginHealthzChecks...)` correct whether we are reloading or not

nilekhc commented:
/cc @liggitt

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go:
```diff
@@ -177,48 +177,48 @@ func TestEncryptionProviderConfigCorrect(t *testing.T) {
+	secretboxFirstTransformer := secretboxFirstEncryptionConfiguration.Transformers[schema.ParseGroupResource("secrets")]
+	kmsFirstTransformer := kmsFirstEncryptionConfiguration.Transformers[schema.ParseGroupResource("secrets")]
+	kmsv2FirstTransformer := kmsv2FirstEncryptionConfiguration.Transformers[schema.ParseGroupResource("secrets")]
 
```
I am testing most math in the `TestKMSMaxTimeout`.  But will also add some assertions in the `TestEncryptionProviderConfigCorrect` to test the `KMSCloseGracePeriod` as well.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -113,25 +115,52 @@ func (h *kmsv2PluginProbe) toHealthzCheck(idx int) healthz.HealthChecker {
+	// The way we calculate is as follows:
+	// 1. calculate max timeout for all KMS plugins
+	// 2. Multiply that by number of KMS plugins
+	// 3. Multiply that by 2 (to consider encryption and decryption call that might be made)
```
For v1 healthz also make calls to Encrypt and Decrypt. I updated the logic accordingly for v1. these comments are no longer valid.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -113,25 +115,52 @@ func (h *kmsv2PluginProbe) toHealthzCheck(idx int) healthz.HealthChecker {
+		Transformers:              transformers,
+		HealthChecks:              kmsHealthChecks,
+		EncryptionFileContentHash: contentHash,
+		KMSTimeout:                time.Duration(2 * 2 * time.Duration(len(kmsHealthChecks)) * kmsUsed.kmsTimeout),
```
updated the logic

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -555,3 +598,116 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+	}
+	if kmsPluginHealthzCheck != nil {
+		dynamicTransformers.kmsPluginHealthzCheckName = kmsPluginHealthzCheck.Name()
+	}
```
true. updated.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -555,3 +598,116 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+// DynamicTransformers holds transformers that are dynamically updated.
+type DynamicTransformers struct {
+	transformTracker          *atomic.Value
+	kmsPluginHealthzCheckName string
```
done

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -555,3 +598,116 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+
+// DynamicTransformers holds transformers that are dynamically updated.
+type DynamicTransformers struct {
+	transformTracker          *atomic.Value
```
I added an issue https://github.com/kubernetes/kubernetes/issues/113759 to follow up on this since we would also need to run some benchmarking. Let me know if this is a blocker.

liggitt replied:
not a blocker, follow-up with evidence is fine

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -555,3 +598,116 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+	dynamicTransformers := &DynamicTransformers{
+		transformTracker: &atomic.Value{},
+	}
+	if kmsPluginHealthzCheck != nil {
```
that's right. updated the logic.

liggitt commented:
thanks, this is much easier to follow now

/approve
/lgtm
/hold for squash of comment commits

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -555,3 +598,116 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+	// close old transformers once we wait for grpc request to finish any in-flight requests.
+	// by the time we spawn this go routine, the new transformers have already been set and will be used for new requests
+	go func() {
+		time.Sleep(oldTransformTracker.kmsTimeout)
```
That is correct. Their life cycle is tied to the server. added a note about it.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -223,12 +228,58 @@ func (s *EtcdOptions) Complete(storageObjectCountTracker flowcontrolrequest.Stor
-		transformerOverrides, kmsPluginHealthzChecks, err := encryptionconfig.LoadEncryptionConfig(s.EncryptionProviderConfigFilepath, s.EncryptionProviderConfigAutomaticReload, stopCh)
+		ctx, closeTransformers := wait.ContextForChannel(stopCh)
+
+		encryptionConfiguration, err := encryptionconfig.LoadEncryptionConfig(s.EncryptionProviderConfigFilepath, s.EncryptionProviderConfigAutomaticReload, ctx.Done())
```
Correct.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,278 @@
+	)
+
+	// get context to close the new transformers.
+	ctx, closeTransformers := wait.ContextForChannel(d.stopCh)
```
correct.

nilekhc commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:
```diff
@@ -0,0 +1,278 @@
+		}
+
+		pollErr := wait.PollImmediate(100*time.Millisecond, 2*time.Minute, func() (bool, error) {
+			healthCheckError := kmsPluginHealthzCheck.Check(req)
```
updated

liggitt commented:
/lgtm

aramase reviewed (commented):
@nilekhc Looks like there is a `hold`. Feel free to remove the hold when ready to merge.

@liggitt @enj I think we need a release note for this?

nilekhc commented:
/unhold

Comments are squashed and got lgtm and approve.

liggitt commented:
the reload flag was already release-noted in https://github.com/kubernetes/kubernetes/pull/113529

liggitt commented:
I updated the description to indicate there is no release note and linked to the PR that added the note
